File: elixir/converter.ex
# To run:
#   iex
#   c("converter.ex")
#   Converter.main "../input.md"
# Pros:
#   - Modules
#   - Atoms, which seem like enums? Maybe
#   - Pattern matching functions
#   - Delta operator
#   - cond is cool
#   - Debugging is easy because everything is printable. I usually knew what was going on through the 'state' of my program
#   - Default values for function parameters
#   - 'when' clause for functions
# Cons:
#   - Functional programming is different than what I'm used to. Not much of a con but it makes it's a new way of solving problems
#   - Can't define functions in other functions (likely because this would mess with overloaded methods. Haskell allows inner functions because it has static function typing)

defmodule Converter do
  defmodule Token, do: defstruct data: "", line: 0, col: 0

  defmodule Header, do: defstruct size: 0, children: []

  defmodule Paragraph, do: defstruct children: []

  defmodule CodeBlock, do: defstruct text: ""

  defmodule Image, do: defstruct text: "", url: ""

  defmodule Text, do: defstruct text: ""

  defmodule Italic, do: defstruct children: []

  defmodule Bold, do: defstruct children: []

  defmodule Code, do: defstruct text: ""

  defmodule Link, do: defstruct text: "", url: ""

  def main(filename) do
    # Open input file, read contents
    contents = File.read!(filename)

    # Create parser, parse documents
    IO.binwrite(File.open!("output.html", [:write]),
    String.graphemes(String.slice(contents, 1, String.length contents))
      |>tokenize(String.first(contents), 1, 1, 1)
      |>parse_document([])
      |>get_HTML_list(""))
  end

  def string_contains(str, c), do: (Enum.member?(String.graphemes(str), c))

  def isSpecialChar(c), do: string_contains("_*`#[]()!", c)

  def tokenize([c | rest], data, line, col, oldCol) do
    oldC = String.last data
    cond do
      string_contains("\r\n", oldC) or
      string_contains("#[(!", c) or
      (isSpecialChar c) != (isSpecialChar oldC) ->  [%Token{data: String.replace(data, "\r", ""), line: line, col: oldCol}] ++ tokenize(rest, c, line, col + 1, col + 1)
      string_contains("\r\n", c) ->                 [%Token{data: String.replace(data, "\r", ""), line: line, col: oldCol}] ++ tokenize(rest, c, line + 1, 0, col + 1)
      true ->                                       tokenize(rest, data <> c, line, col + 1, oldCol)
    end
  end

  def tokenize([], data, line, col, _) do
    [%Token{data: data, line: line, col: col}, %Token{data: "\n", line: line, col: col}]
  end

  def syntax_error(token, msg) do
    IO.puts "error: #{token.line}:#{token.col} #{msg}, got `#{token.data}`"
    Process.exit(self, :normal)
  end

  def getHeaderSize([head | tail], acc \\ 1) do
    if head.data == "#" do
      getHeaderSize(tail, acc + 1)
    else
      {acc, [head] ++ tail}
    end
  end

  # -> (list of nodes, rest of tokens)
  def parse_formatted_text([top | rest], bounds, acc \\ []) do
    cond do
      Enum.member?(bounds, top.data)  -> {acc, rest}
      top.data === "*" or
      top.data === "_" -> (
        {children, new_rest} = parse_formatted_text(rest, ["*", "_"] ++ bounds)
        parse_formatted_text(new_rest, bounds, acc ++ [%Italic{children: children}])
      )
      top.data === "**" or
      top.data === "__" -> (
        {children, new_rest} = parse_formatted_text(rest, ["**", "__"] ++ bounds, [])
        parse_formatted_text(new_rest, bounds, acc ++ [%Bold{children: children}])
      )
      top.data === "`" -> (
        {node, new_rest} = parse_code(rest)
        parse_formatted_text(new_rest, bounds, acc ++ [node])
      )
      top.data === "[" -> (
        {node, new_rest} = parse_link(rest)
        parse_formatted_text(new_rest, bounds, acc ++ [node])
      )
      true -> parse_formatted_text(rest, bounds, acc ++ [%Text{text: top.data}])
    end
  end

  def take_until([top | rest], sentinel, acc) do
    if top.data == sentinel do
      {acc, rest}
    else
      take_until(rest, sentinel, acc <> top.data)
    end
  end

  def parse_document(tokens, acc) when is_list(tokens) and length(tokens) > 0 do
    {node, rest} = parse_node tokens
    if node do
      parse_document(rest, acc ++ [node])
    else
      parse_document(rest, acc)
    end
  end

  def parse_document([], acc), do: acc

  def parse_node([top | rest]) do
    case top.data do
      "#"   -> parse_header(rest)
      "```" -> parse_codeblock(rest)
      "!"   -> parse_image(rest)
      "\n"  -> {nil, rest}
      _     -> parse_paragraph([top] ++ rest)
    end
  end

  def parse_header(tokens) do
    {size, rest} = getHeaderSize tokens
    {children, rest} = parse_formatted_text(rest, ["\n"])
    {%Header{size: size, children: children}, rest}
  end

  def parse_paragraph(tokens) do
    {children, rest} = parse_formatted_text(tokens, ["\n"])
    {%Paragraph{children: children}, rest}
  end

  def parse_codeblock(tokens) do
    {text, rest} = take_until(tokens, "```", "")
    {%CodeBlock{text: text}, rest}
  end

  def parse_image([%Token{data: "["}, %Token{data: text}, %Token{data: "]"}, %Token{data: "("}, %Token{data: url}, %Token{data: ")"} | rest]) do
    {%Image{text: text, url: url}, rest}
  end

  def parse_code(tokens) do
    {text, rest} = take_until(tokens, "`", "")
    {%Code{text: text}, rest}
  end

  def parse_link([%Token{data: text}, %Token{data: "]"}, %Token{data: "("}, %Token{data: url}, %Token{data: ")"} | rest]) do
    {%Link{text: text, url: url}, rest}
  end

  def parse_link([top | _]) do
    syntax_error(top, "expected a link")
  end

  def get_HTML_list_header([top | rest]) do
    get_HTML_list(rest, String.trim(get_HTML(top)))
  end

  def get_HTML_list([top | rest], acc) do
    get_HTML_list(rest, acc <> get_HTML(top))
  end

  def get_HTML_list([], acc), do: acc

  def get_HTML(node) do
    case node do
      %Header{size: size, children: children} -> "<h#{Integer.to_string(size)}>#{get_HTML_list_header(children)}</h#{Integer.to_string(size)}>\n"
      %Paragraph{children: children}          -> "<p>#{get_HTML_list(children, "")}</p>\n\n"
      %CodeBlock{text: text}                  -> "<pre><code>#{text}</code></pre>\n\n"
      %Image{text: text, url: url}            -> "<img src=\"#{url}\" alt=\"#{text}\" />"
      %Text{text: text}                       -> text
      %Italic{children: children}             -> "<em>#{get_HTML_list(children, "")}</em>"
      %Bold{children: children}               -> "<strong>#{get_HTML_list(children, "")}</strong>"
      %Code{text: text}                       -> "<code>#{text}</code>"
      %Link{text: text, url: url}             -> "<a href=\"#{url}\">#{text}</a>"
    end
  end
end

File: ocaml/converter.ml
(* To run: ocamlc converter.ml && ./a.out
Pros:
  - Nice record syntax/typing
  - Pattern matching
Cons:
  - No line comments
  - `let in` everywhere
  - Really poor divergence
  - No generic print function
  - `else-if`
  - Poor standard library support (but maybe Lists aren't really the best option in OCaml)
  - Non-mutally recursive functions (fairly lame!)
*)

let explode s = List.init (String.length s) (String.get s)
let implode l = String.of_seq (List.to_seq l)

let rec last = function
  | x::[] -> x
  | _::xs -> last xs
  | []    -> failwith "no element"

let read_whole_file filename =
  (* open_in_bin works correctly on Unix and Windows *)
  let ch = open_in_bin filename in
  let s = really_input_string ch (in_channel_length ch) in
  close_in ch;
  s

type token = {data: string; line: int; col: int}

type node = 
  | Header of int * node list
  | Paragraph of node list
  | CodeBlock of string
  | Image of string * string
  | Text of string
  | Italic of node list
  | Bold of node list
  | Code of string
  | Link of string * string

let is_special_char c = String.contains "_*`#[]()!" c

let rec tokenize contents token line col oldCol = 
  match contents with
  | []      -> [{data = (implode token); line = line; col = col}]
  | c::rest -> 
    let oldC = last(token) in
    if c == '\n' 
      then {data = (implode token); line = line; col = oldCol} :: tokenize rest [c] (line + 1) 0 (col + 1)
    else if oldC == '\n' || (String.contains "#[(!" c) || (is_special_char c != is_special_char oldC) 
      then {data = (implode token); line = line; col = oldCol} :: tokenize rest [c] (line + 1) 0 (col + 1)
    else tokenize rest (token @ [c]) line (col + 1) oldCol

let rec get_header_size tokens acc =
  match tokens with
  | [] -> failwith "Expected a header."
  | head::tail -> 
    if String.equal head.data "#" then
      get_header_size tail (acc + 1)
    else
      acc, head::tail

let rec take_until tokens sentinel acc =
  match tokens with
  | [] -> failwith "Expected sentinel."
  | top::rest ->
    if String.equal top.data sentinel then
      (acc, rest)
    else
      take_until rest sentinel (acc ^ top.data)

let parse_code tokens =
  let (text, rest) = take_until tokens "`" "" in
  (Code text), rest

let parse_link tokens =
  match tokens with
  | {data = text; _}::{data = "]"; _}::{data = "("; _}::{data = url; _}::{data = ")"; _}::rest -> 
    Link(text, url), rest
  | _ -> failwith "Expected link text."

let rec parse_formatted_text tokens bounds acc =
  match tokens with
  | [] -> failwith "Expected formatted text."
  | top::rest ->
    if List.mem (top.data) bounds then 
      acc, rest
    else if (String.equal top.data "*") || (String.equal top.data "_") then
      let (children, new_rest) = parse_formatted_text rest ("*"::"_"::bounds) [] in
      parse_formatted_text new_rest bounds (acc @ [Italic children])
    else if (String.equal top.data "**") || (String.equal top.data "__") then
      let (children, new_rest) = parse_formatted_text rest ("**"::"__"::bounds) [] in
      parse_formatted_text new_rest bounds (acc @ [Bold children])
    else if (String.equal top.data "`") then
      let (node, new_rest) = parse_code rest in
      parse_formatted_text new_rest bounds (acc @ [node])
    else if (String.equal top.data "[") then
      let (node, new_rest) = parse_link rest in
      parse_formatted_text new_rest bounds (acc @ [node])
    else
      parse_formatted_text rest bounds (acc @[Text top.data])

let parse_header tokens =
  let (size, rest) = get_header_size tokens 1 in
  let (children, rest2) = parse_formatted_text rest ["\n"] [] in
  Some (Header(size, children)), rest2

let parse_paragraph tokens =
  let (children, rest) = parse_formatted_text tokens ["\n"] [] in
  Some (Paragraph(children)), rest

let parse_codeblock tokens =
  let (text, rest) = take_until tokens "```" "" in
  Some (CodeBlock(text)), rest

let parse_image tokens =
  match tokens with
  | {data = "["; _}::{data = text; _}::{data = "]"; _}::{data = "("; _}::{data = url; _}::{data = ")"; _}::rest -> 
    Some (Image(text, url)), rest
  | _ -> failwith "Expected link text."

let parse_node contents =
  match contents with
  | []        -> (None, [])
  | top::rest -> 
    match top.data with
    | "#"   -> parse_header rest
    | "```" -> parse_codeblock rest
    | "!"   -> parse_image rest
    | "\n"  -> (None, rest)
    | _     -> parse_paragraph ([top] @ rest)

let rec parse_document tokens acc =
  match tokens with
  | [] -> acc
  | _  ->
    let (maybe_node, rest) = parse_node tokens in
    match maybe_node with
    | None      -> parse_document rest acc
    | Some node -> parse_document rest (acc @ [node])

let rec get_HTML node =
  match node with
  | Header (size, children) -> 
    let child = get_HTML_list_header children in
    Printf.sprintf "<h%d>%s</h%d>\n" size child size
  | Paragraph children -> Printf.sprintf "<p>%s</p>\n\n" (get_HTML_list children "")
  | CodeBlock text -> Printf.sprintf "<pre><code>%s</code></pre>\n\n" text
  | Image (text, url) -> Printf.sprintf "<img src=\"%s\" alt=\"%s\" />" url text
  | Text text -> text
  | Italic children -> Printf.sprintf "<em>%s</em>" (get_HTML_list children "")
  | Bold children -> Printf.sprintf "<strong>%s</strong>" (get_HTML_list children "")
  | Code text -> Printf.sprintf "<code>%s</code>" text
  | Link (text, url) -> Printf.sprintf "<a href=\"%s\">%s</a>" url text

and get_HTML_list tokens acc =
  match tokens with
  | [] -> acc
  | top::rest -> get_HTML_list rest (acc ^ (get_HTML top))

and get_HTML_list_header tokens =
  match tokens with
  | [] -> failwith "Expected something in the header, tbh"
  | top::rest -> get_HTML_list rest (String.trim (get_HTML top))

let () =
  let input = explode (read_whole_file "../input.md") in
  match input with
  | [] -> ()
  | first::contents ->
    let tokens = tokenize contents [first] 1 1 1 in
    let document = parse_document tokens [] in
    let output_str = get_HTML_list document "" in
    let oc = open_out "output.html" in
    Printf.fprintf oc "%s\n" output_str;
    close_out oc;